### PR TITLE
Added BitBucket SCM provider

### PR DESCRIPTION
**Pull request by RobertNorthard**
**As** a developer
**I want** a BitBucket SCM provider.

### CODE REVIEW

**Comment by dsingh07** [comment]
@RobertNorthard Can we squash the commit which says "Removed not working part"?

**Comment by RobertNorthard** [comment]
@dsingh07 rebased 👍

File: src/main/groovy/pluggable/scm/bitbucket/BitbucketRequestUtil.groovy [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    HttpURLConnection http = (HttpURLConnection) url.openConnection();
+    http.setRequestMethod("GET");
+    http.setRequestProperty ("Authorization", "Basic ${auth}");
+    http.setRequestProperty("Content-Type", "application/json");
+
+    switch (http.getResponseCode()) {
+      case 200:
+        def json = jsonSlurper.parse(http.getInputStream())
+        for(int i = 0; i < json.size; i++){
+          repositoryList.add(json.values[i].name)
```
Comment (SachinKSingh28) [inline src/main/groovy/pluggable/scm/bitbucket/BitbucketRequestUtil.groovy:52]: we should check for the paging in the json response as well.

File: src/main/groovy/pluggable/scm/bitbucket/BitbucketSCMProvider.groovy [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+            default:
+              throw new IllegalArgumentException("SCM Protocol type not supported.");
+              break;
+          }
+
+          url.append(this.bitbucketHost);
+          url.append(":");
+          url.append(this.scmPort);
+          url.append(this.bitbucketEndpoint);
+          if(this.bitbucketEndpoint != "/"){
```
Comment (SachinKSingh28) [inline src/main/groovy/pluggable/scm/bitbucket/BitbucketSCMProvider.groovy]: should check for a trailing "/" and append if it doesn't exists.

**Review by SachinKSingh28 — CHANGES_REQUESTED** [review]

File: src/main/groovy/pluggable/scm/bitbucket/BitbucketRequestUtil.groovy (hunk 1 above)
Comment (RobertNorthard) [reply src/main/groovy/pluggable/scm/bitbucket/BitbucketRequestUtil.groovy:52]: Good shout! I've defaulted the limit to 100. We will add this as an issue to support pagination.

File: src/main/groovy/pluggable/scm/bitbucket/BitbucketSCMProvider.groovy [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+
+        def gitSsh = new File (tempDir + '/git_ssh.sh')
+        def tempScript = new File(tempDir + '/shell_script.sh')
+
+        gitSsh << "#!/bin/sh\n"
+        gitSsh << "exec ssh -i " + envVarProperty.getSshPrivateKeyPath() + " -o StrictHostKeyChecking=no \"\$@\""
+
+        tempScript << "export GIT_SSH=\""+ tempDir + "/git_ssh.sh\"\n"
+        tempScript << "git clone " + this.getScmUrl(this.bitbucketUsername, this.bitbucketPassword) + repoNamespace + "/" + repoName + ".git " + tempDir + "/" + repoName + "\n"
+        def gitDir = "--git-dir=" + tempDir + "/" + repoName + "/.git"
```
Comment (RobertNorthard) [inline src/main/groovy/pluggable/scm/bitbucket/BitbucketSCMProvider.groovy]: BitBucket password should be ASCII encoded as it's a url. What if someone has an '@' in their password?

**Review by dsingh07 — COMMENTED** [review]
Looks good @RobertNorthard @SachinKSingh28 

Have tested this one a fresh /C and it works as expected, in conjunction with https://github.com/Accenture/adop-platform-management/pull/30

Once the commits have been squashed, this is ready to be merged in.

**Comment by dsingh07** [comment]
LGTM, thanks @RobertNorthard @BuleGeorge @SachinKSingh28 @MarisBahtins @IrmantasM

**Comment by SachinKSingh28** [comment]
I think we need to publish the github pages as well.

**Comment by RobertNorthard** [comment]
Awesome, thanks @SachinKSingh28. I've updated GitHub pages. The latest changes are now being reflected (https://accenture.github.io/adop-pluggable-scm/javadocs/)

**Pull request merged**